hnn_core/parallel_backends.py

```diff
@@ -554,34 +555,237 @@ def simulate(self, net, tstop, dt, n_trials, postproc=False):
-              f"distributing trials over {self.n_jobs} jobs.")
+        print(
+            f"Joblib will run {n_trials} trial(s) in parallel by "
+            f"distributing trials over {self.n_jobs} jobs."
+        )
         parallel, myfunc = self._parallel_func(_simulate_single_trial)
-        sim_data = parallel(myfunc(net, tstop, dt, trial_idx) for
-                            trial_idx in range(n_trials))
+        sim_data = parallel(
+            myfunc(net, tstop, dt, trial_idx) for trial_idx in range(n_trials)
+        )
 
-        dpls = _gather_trial_data(sim_data, net=net, n_trials=n_trials,
-                                  postproc=postproc)
+        dpls = _gather_trial_data(
+            sim_data, net=net, n_trials=n_trials, postproc=postproc
+        )
 
         return dpls
 
 
+def _determine_cores_hwthreading(
+        enable_hwthreading: Union[None, bool] = True,
+        sensible_default_cores: bool = False,
+) -> [int, bool]:
+    """Return the number of available cores and if hardware-threading is used.
+
+    This is important for systems where the number of available cores is
+    partitioned such as on HPC systems, but is also important for determining
+    hardware support for hardware-threading. Hardware-threading ("hwthread" in
+    OpenMPI parlance), simultaneous multi-threading (SMT,
+    https://en.wikipedia.org/wiki/Simultaneous_multithreading ), and [Intel]
+    Hyperthreading are all terms used interchangeably, and are essentially
+    equivalent.
+
+    Parameters
+    ----------
+    enable_hwthreading : bool
+        Whether to detect support for hardware-threading and, if the feature is
+        detected, return the available number of 'logical' hardware-threaded
+        cores. Defaults to True. If 'False', or the feature is not detected,
+        return the available number of 'physical' cores (excluding
+        double-counting of hardware-threaded cores).
+    sensible_default_cores : bool
+        Whether to decrease the number of cores returned in a reasonable
+        manner, such that it balances speed with the user experience (e.g.,
+        preventing the machine 'locking-up'). Defaults to 'False'.
+
+    Returns
+    -------
+    core_count : int
+        Number of logical CPU cores available for use by a process.
+    hwthreading_present : bool
+        Whether or not hardware-threading is present on some or all of the
+        logical CPU cores.
+    """
+    # Needs its own import checks since it may be called by the GUI before
+    # MPIBackend()
+    if _has_mpi4py() and _has_psutil():
+        if enable_hwthreading is None:
+            # This lets us pass the same arg to this function and MPIBackend()
+            # in case we want to use the default approaches.
+            enable_hwthreading = True
+        import platform
+        import psutil
+        if platform.system() == "Darwin":
+            # In Macos' case here, the situation is relatively simple, and we
```

Comment:
It honestly seems like the ability to reliably detect the number of cores could be a standalone python package